en/tutorials-and-examples/cms/authentication-plugin.rst

```diff
@@ -0,0 +1,288 @@
+to generate the classes for you using ``bin/cake bake all users``.
+
+Adding Password Hashing
+-----------------------
+
+If you were to create/update a user at this point in time, you might notice that
+the passwords are stored in plain text. This is really bad from a security point
+of view, so lets fix that.
+
+This is also a good time to talk about the model layer in CakePHP. In CakePHP,
+we separate the methods that operate on a collection of objects, and a single
+object into different classes. Methods that operate on the collection of
+entities are put in the ``Table`` class, while features belonging to a single
+record are put on the ``Entity`` class.
+
+For example, password hashing is done on the individual record, so we'll
+implement this behavior on the entity object. Because we want to hash the
+password each time it is set, we'll use a mutator/setter method. CakePHP will
+call convention based setter methods any time a property is set in one of your
+entities. Let's add a setter for the password. In **src/Model/Entity/User.php**
+add the following::
+
+    <?php
+    namespace App\Model\Entity;
+
+    use Cake\Auth\DefaultPasswordHasher; // Add this line
+    use Cake\ORM\Entity;
+
+    class User extends Entity
+    {
+
+        // Code from bake.
+
+        // Add this method
+        protected function _setPassword(string $password) : ?string
+        {
+            if (strlen($password) > 0) {
+                return (new DefaultPasswordHasher())->hash($password);
+            }
+        }
+    }
+
+Now, point your browser to **http://localhost:8765/users** to see a list of users.
+Remember you'll need to have your local server running, start a standalone PHP
+server using ``bin/cake server``
+
+You can edit the default user that was created during
+:doc:`Installation <installation>`. If you change that user's password,
+you should see a hashed password instead of the original value on the list or
+view pages. CakePHP hashes passwords with `bcrypt
+<http://codahale.com/how-to-safely-store-a-password/>`_ by default. We recommend
+bcrypt for all new applications to keep your security standards high.
+
+.. note::
+
+    Create a hashed password for at least one of the user accounts now! 
+    It will be needed in the next steps.
+    After updating the password, you'll see a long string stored in the password column.
+    Note bcrypt will generate a different hash even for the same password saved twice.
+    
+
+Installing Authentication Plugin
+================================
+
+Use composer to install the Authentication Plugin
+
+    composer require cakephp/authentication:^2.0
+
+Adding Login
+============
+
+Now it's time to configure the Authentication Plugin.
+The Plugin will handle the authentication process using 3 different classes:
+
+* ``Application`` will use the Authentication Middleware and provide an
+AuthenticationService, holding all the configuration we want to define how are
```

Comment:
```suggestion
  AuthenticationService, holding all the configuration we want to define how are
```

The wrapped lines of a bulleted list need to be aligned to be formatted correctly.